test/browser/render.test.js

```diff
@@ -789,6 +791,50 @@ describe('render()', () => {
 		expect(checkbox.checked).to.equal(true);
 	});
 
+	it('should always diff `contenteditable` `innerHTML` against the DOM', () => {
+		// This tests that we do not cause any cursor jumps in contenteditable fields
+		// See https://github.com/preactjs/preact/issues/2691
+
+		function Editable() {
+			const [value, setValue] = useState('Hello');
+
+			return (
+				<div
+					contentEditable
+					dangerouslySetInnerHTML={{ __html: value }}
+					onInput={e => setValue(e.currentTarget.innerHTML)}
+				/>
+			);
+		}
+
+		render(<Editable />, scratch);
+
+		// select all the text inside the contenteditable div
+		let editable = scratch.querySelector('[contenteditable]');
+		const range = document.createRange();
+		range.selectNodeContents(editable);
+		const sel = window.getSelection();
+		sel.removeAllRanges();
+		sel.addRange(range);
+
+		expect(window.getSelection().getRangeAt(0).endOffset).to.equal(1);
+
+		// dispatch a change event
+		editable.innerHTML = 'World';
+		editable.dispatchEvent(createEvent('input'));
```

Comment:
Hmm I don't think I understand this test. How can a real user fire an input event or change the innerHTML but simultaneously keep text selection. My understanding is that all browsers would clear the text selection the moment the user presses a key while selecting text.

Reply:
Yeah you're right. I forgot to get back to this. I originally wanted to use the range to place the caret (see https://stackoverflow.com/a/6249440/2164347) but failed. I will try again to make the test clear. In the end this just makes sure we do not touch innerHTML.

Reply:
I modified the test to be more realistic. It now ensures the caret stays where it was prior to the `input` event